File: apps/server/src/db/url.ts
// Connection URL parsing: `postgres://`, `postgresql://`, `pg://`, `mysql://`, `mariadb://`.

import os from "node:os";
import { type ConnectionConfig, type Dialect } from "@perch/protocol";

export const defaultPort = (dialect: Dialect): number => (dialect === "mysql" ? 3306 : 5432);

const PROTOCOLS: Record<string, Dialect> = {
  "postgres:": "postgres",
  "postgresql:": "postgres",
  "pg:": "postgres",
  "mysql:": "mysql",
  "mariadb:": "mysql",
};

/** Query parameters that configure the connection rather than the driver options bag. */
const RESERVED_PARAMS = new Set(["ssl", "sslmode", "dialect"]);

function decode(value: string): string {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
}

function parseSsl(params: URLSearchParams): boolean | undefined {
  const sslmode = params.get("sslmode");
  if (sslmode) {
    const mode = sslmode.toLowerCase();
    if (mode === "disable" || mode === "allow") return false;
    return true;
  }
  const ssl = params.get("ssl");
  if (ssl === null) return undefined;
  const value = ssl.toLowerCase();
  if (value === "" || value === "1" || value === "true" || value === "require") return true;
  if (value === "0" || value === "false") return false;
  return true;
}

function osUser(): string {
  try {
    return os.userInfo().username;
  } catch {
    return "postgres";
  }
}

/**
 * Parses `postgres://`, `postgresql://` and `mysql://` URLs into a connection config.
 * Missing pieces fall back to sensible defaults (the OS user and `postgres` for PostgreSQL,
 * `root` for MySQL). `?sslmode=require` / `?ssl=true` turn TLS on; any other query parameter is
 * kept in `options` for the driver.
 */
export function parseConnectionUrl(
  url: string,
): Omit<ConnectionConfig, "id" | "name" | "createdAt"> {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error(`invalid connection url: ${url}`);
  }

  const dialect = PROTOCOLS[parsed.protocol.toLowerCase()];
  if (!dialect) {
    throw new Error(
      `unsupported connection url scheme "${parsed.protocol.replace(/:$/, "")}" — use postgres:// or mysql://`,
    );
  }

  const host = decode(parsed.hostname).replace(/^\[|\]$/g, "") || "localhost";
  const port = parsed.port ? Number(parsed.port) : defaultPort(dialect);
  if (!Number.isInteger(port) || port <= 0 || port > 65535) {
    throw new Error(`invalid port in connection url: ${parsed.port}`);
  }

  const user = decode(parsed.username) || (dialect === "mysql" ? "root" : osUser());
  const password = parsed.password ? decode(parsed.password) : undefined;
  const database =
    decode(parsed.pathname.replace(/^\//, "")) || (dialect === "postgres" ? "postgres" : "");

  const ssl = parseSsl(parsed.searchParams);
  const options: Record<string, string> = {};
  for (const [key, value] of parsed.searchParams) {
    if (!RESERVED_PARAMS.has(key.toLowerCase())) options[key] = value;
  }

  const config: Omit<ConnectionConfig, "id" | "name" | "createdAt"> = {
    dialect,
    host,
    port,
    user,
    database,
  };
  if (password !== undefined) config.password = password;
  if (ssl !== undefined) config.ssl = ssl;
  if (Object.keys(options).length > 0) config.options = options;
  return config;
}

File: apps/server/db/url.go
package db

import (
	"fmt"
	"net/url"
	"os/user"
	"strconv"
	"strings"

	"perch/protocol"
)

func DefaultPort(dialect protocol.Dialect) int {
	if dialect == protocol.DialectMySQL {
		return 3306
	}
	return 5432
}

var urlSchemes = map[string]protocol.Dialect{
	"postgres":   protocol.DialectPostgres,
	"postgresql": protocol.DialectPostgres,
	"pg":         protocol.DialectPostgres,
	"mysql":      protocol.DialectMySQL,
	"mariadb":    protocol.DialectMySQL,
}

// Query parameters that configure the connection rather than the driver options bag. `sslmode`
// is not among them: it sets the SSL flag *and* stays in Options, because "on" is not the whole
// answer — a pasted hosted-Postgres URL says which of the five modes it wants, and dropping that
// silently downgrades verify-full to an unverified tunnel.
var reservedParams = map[string]bool{"ssl": true, "dialect": true}

// NewDriver builds the driver for a connection's dialect.
func NewDriver(config protocol.ConnectionConfig) (Driver, error) {
	switch config.Dialect {
	case protocol.DialectPostgres:
		return newPostgresDriver(config), nil
	case protocol.DialectMySQL:
		return newMysqlDriver(config), nil
	}
	return nil, fmt.Errorf("unsupported dialect: %s", config.Dialect)
}

// ParseConnectionURL turns a postgres:// or mysql:// URL into a connection config. Missing
// pieces fall back to the OS user and `postgres` for PostgreSQL, `root` for MySQL;
// ?sslmode=require or ?ssl=true turns TLS on, and any other query parameter is kept in Options.
func ParseConnectionURL(raw string) (protocol.ConnectionConfig, error) {
	var config protocol.ConnectionConfig
	parsed, err := url.Parse(raw)
	if err != nil {
		return config, fmt.Errorf("invalid connection url: %s", raw)
	}
	dialect, ok := urlSchemes[strings.ToLower(parsed.Scheme)]
	if !ok {
		return config, fmt.Errorf(
			"unsupported connection url scheme %q — use postgres:// or mysql://", parsed.Scheme)
	}

	host := strings.Trim(parsed.Hostname(), "[]")
	if host == "" {
		host = "localhost"
	}
	port := DefaultPort(dialect)
	if raw := parsed.Port(); raw != "" {
		port, err = strconv.Atoi(raw)
		if err != nil || port <= 0 || port > 65535 {
			return config, fmt.Errorf("invalid port in connection url: %s", parsed.Port())
		}
	}

	user := ""
	password := ""
	if parsed.User != nil {
		user = parsed.User.Username()
		password, _ = parsed.User.Password()
	}
	if user == "" {
		if dialect == protocol.DialectMySQL {
			user = "root"
		} else {
			user = osUser()
		}
	}

	database := strings.TrimPrefix(parsed.Path, "/")
	if database == "" && dialect == protocol.DialectPostgres {
		database = "postgres"
	}

	query := parsed.Query()
	options := map[string]any{}
	for key, values := range query {
		if !reservedParams[strings.ToLower(key)] && len(values) > 0 {
			options[key] = values[0]
		}
	}

	config = protocol.ConnectionConfig{
		Dialect:  dialect,
		Host:     host,
		Port:     port,
		User:     user,
		Password: password,
		Database: database,
		SSL:      parseSSL(query),
	}
	if len(options) > 0 {
		config.Options = options
	}
	return config, nil
}

func parseSSL(query url.Values) bool {
	if mode := query.Get("sslmode"); mode != "" {
		switch strings.ToLower(mode) {
		case "disable", "allow":
			return false
		}
		return true
	}
	if !query.Has("ssl") {
		return false
	}
	switch strings.ToLower(query.Get("ssl")) {
	case "0", "false":
		return false
	}
	return true
}

func osUser() string {
	if current, err := user.Current(); err == nil && current.Username != "" {
		return current.Username
	}
	return "postgres"
}
